### PR TITLE
`<execution>`, `<numeric>`: Change improper direct-initialization to copy-initialization in C++17 numeric algorithms

### PR DESCRIPTION
**Pull request by frederick-vs-ja**
Follows up #4260.

Affected algorithms:
- `reduce`
- `transform_reduce`
- `exclusive_scan`
- `inclusive_scan`
- `transform_exclusive_scan`
- `transform_inclusive_scan`

The standard wording only uses "convertible" (from the results of operations to the element type) in the requirements of these algorithms, so direct-initialization can be sometimes wrong.

The construction from `_Transform_op(*_First)` seems incorrect for _`meow`_`scan` algorithms as the standard doesn't seem to require this. See also #3783. Perhaps we should remove such construction later.

`_Implicitly_construct_in_place_`_`meow`_ function templates are added for performing implicit conversion in placements new and avoiding additional move, which may be important for WG21-P2408R5.

### CODE REVIEW

**Comment by frederick-vs-ja** [comment]
😿 Sorry again for force-pushing due to misconfiguration...

File: stl/inc/execution [hunk 1]
```diff
@@ -137,6 +137,35 @@ template <>
...
+    ::new (static_cast<void*>(_STD addressof(_Val))) _Ty(*_Iter);
+}
+
+template <class _Ty, class _UnaryOp, class _FwdIt>
+void _Construct_in_place_by_transform_deref(_Ty& _Val, _UnaryOp& _Tranform_op, const _FwdIt& _Iter) {
+    ::new (static_cast<void*>(_STD addressof(_Val))) _Ty(_Tranform_op(*_Iter));
+}
+
+template <class _Ty, class _BinaryOp, class _ArgTy>
+void _Implicitly_construct_in_place_by_binary_op(_Ty& _Val, _BinaryOp& _Reduce_op, _ArgTy& _Left, _ArgTy& _Right) {
```
Comment (AlexGuteniev) [inline stl/inc/execution:151]: Passing op by reference here is inconsistent with passing by value in the calling function.

File: stl/inc/execution [hunk 2]
```diff
@@ -137,6 +137,35 @@ template <>
...
+    ::new (static_cast<void*>(_STD addressof(_Val))) _Ty([&]() -> _Ty { return _Reduce_op(_Left, _Right); }());
```
Comment (AlexGuteniev) [inline stl/inc/execution:152]: And capturing it by reference too. Ditto below.

Comment (frederick-vs-ja) [reply stl/inc/execution:152]: The point is performing implicit conversion in placement new with workaround for imperfectness of forwarding. So I guess copying should be avoided.

File: tests/std/tests/GH_004129_conversion_in_new_numeric_algorithms/test.cpp [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+    operator int64_t() const noexcept {
+        return n;
+    }
+};
+
+static_assert(!is_constructible_v<int32_t, implicitly_convertible_to_i32_only>);
+static_assert(is_convertible_v<implicitly_convertible_to_i32_only, int32_t>);
+
+struct implicity_validating_converter {
```
Comment (StephanTLavavej) [inline tests/std/tests/GH_004129_conversion_in_new_numeric_algorithms/test.cpp:30]: Repeated typo: "implicity"

File: tests/std/tests/GH_004129_conversion_in_new_numeric_algorithms/test.cpp [hunk 4]
```diff
@@ -0,0 +1,185 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// intentionally test narrowing conversion form int64_t to int32_t
```
Comment (StephanTLavavej) [inline tests/std/tests/GH_004129_conversion_in_new_numeric_algorithms/test.cpp:4]: Typo: form

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for fixing these algorithms! :hammer_and_wrench: :abacus: :magic_wand:

**Pull request merged**